paddle/fluid/memory/allocation/allocator_facade.cc

```diff
@@ -399,42 +413,75 @@ class AllocatorFacadePrivate {
     }
   }
 
-  const std::shared_ptr<StreamSafeCUDAAllocator>
-  GetDefaultStreamSafeCUDAAllocator(const platform::CUDAPlace& place) const {
-    const auto iter = default_stream_safe_cuda_allocators_.find(place);
-    PADDLE_ENFORCE_NE(
-        iter,
-        default_stream_safe_cuda_allocators_.end(),
-        platform::errors::NotFound(
-            "No StreamSafeCUDAAllocator found for the place, %s", place));
-    return iter->second;
+  const std::shared_ptr<Allocator> GetDefaultStreamSafeCUDAAllocator(
+      const platform::CUDAPlace& place) const {
+    if (auto iter = default_stream_safe_cuda_allocators_.find(place);
+        iter != default_stream_safe_cuda_allocators_.end())
+      return iter->second;
+    if (auto iter = default_cuda_malloc_async_allocators_.find(place);
+        iter != default_cuda_malloc_async_allocators_.end())
+      return iter->second;
+    PADDLE_THROW(platform::errors::NotFound(
+        "No StreamSafeCUDAAllocator found for the place, %s", place));
   }
 
   gpuStream_t GetDefaultStream(const platform::CUDAPlace& place) const {
-    const std::shared_ptr<StreamSafeCUDAAllocator>& allocator =
-        GetDefaultStreamSafeCUDAAllocator(place);
-    return allocator->GetDefaultStream();
+    if (auto allocator = std::dynamic_pointer_cast<StreamSafeCUDAAllocator>(
+            GetDefaultStreamSafeCUDAAllocator(place))) {
+      return allocator->GetDefaultStream();
+    } else if (auto allocator =
+                   std::dynamic_pointer_cast<CUDAMallocAsyncAllocator>(
+                       GetDefaultStreamSafeCUDAAllocator(place))) {
+      return allocator->GetDefaultStream();
+    } else {
+      PADDLE_THROW(platform::errors::NotFound(
+          "No StreamSafeCUDAAllocator or CUDAMallocAsyncAllocator found for "
+          "the place, %s",
+          place));
+    }
   }
 
   void SetDefaultStream(const platform::CUDAPlace& place, gpuStream_t stream) {
-    const std::shared_ptr<StreamSafeCUDAAllocator>& allocator =
-        GetDefaultStreamSafeCUDAAllocator(place);
-
-    PADDLE_ENFORCE_EQ(
-        allocator->GetDefaultStream(),
-        nullptr,
-        platform::errors::Unavailable(
-            "The default stream for StreamSafeCUDAAllocator(%p) in %s has been "
-            "set to %p, not allow to change it to %p.",
-            allocator.get(),
-            place,
-            allocator->GetDefaultStream(),
-            stream));
-
-    allocator->SetDefaultStream(stream);
-    VLOG(8) << "Set default stream to " << stream
-            << " for StreamSafeCUDAAllocator(" << allocator.get() << ") in "
-            << place;
+    if (auto allocator = std::dynamic_pointer_cast<StreamSafeCUDAAllocator>(
+            GetDefaultStreamSafeCUDAAllocator(place))) {
```

Comment:
We have 3 design choices here


1. Inheritance from Allocator: Currently, both StreamSafeCUDAAllocator and CUDAMallocAsyncAllocator inherit directly from the Allocator class. However, only StreamSafeCUDAAllocator and CUDAMallocAsyncAllocator possess stream-related methods, while other allocator types do not.
   * Pros: This approach aligns well with the conceptual model where CUDAMallocAsyncAllocator is seen as a alternative to StreamSafeCUDAAllocator.
   * Cons: We must upcast to Allocator and then downcast back to CUDAMallocAsyncAllocator or StreamSafeCUDAAllocator.
2. Centralizing Stream-Related Methods: Another approach is to move stream-related methods into the base Allocator class. For allocators that do not support these methods, they would trigger a runtime error.
   * Cons: It complicates the base class with methods that are irrelevant for some of its subclasses, violating the principle of interface segregation.
3. Inheriting CUDAMallocAsyncAllocator from StreamSafeCUDAAllocator: The third option considers making CUDAMallocAsyncAllocator inherit from StreamSafeCUDAAllocator. This approach implies a direct relationship between the two, with one being a more specific version of the other.
   * Cons: This design is conceptually awkward as it positions CUDAMallocAsyncAllocator as a subtype of StreamSafeCUDAAllocator, despite it being intended as a replacement. It might imply a false relationship or hierarchy, leading to confusion and potentially misused inheritance. 

![Paddle drawio (11)](https://github.com/PaddlePaddle/Paddle/assets/23652381/570eb984-be93-48d3-8a60-1a3b4016dced)